src/sas/sascalc/fit/fitstate.py

```diff
@@ -0,0 +1,407 @@
+"""
+Interface to page state for saved fits.
+
+The 4.x sasview gui builds the model, smearer, etc. directly inside the wx
+GUI code.  This code separates the in-memory representation from the GUI.
+Initially it is used for a headless bumps fitting backend that operates
+directly on the saved XML; eventually it should provide methods to replace
+direct access to the PageState object so that the code for setting up and
+running fits in wx, qt, and headless versions of SasView shares a common
+in-memory representation.
+"""
+from __future__ import print_function, division
+
+import copy
+from collections import namedtuple
+
+import numpy as np
+
+from bumps.names import FitProblem
+
+from sasmodels.core import load_model_info
+from sasmodels.data import plot_theory
+from sasmodels.sasview_model import _make_standard_model, MultiplicationModel, load_custom_model
+from sasmodels.weights import MODELS as POLYDISPERSITY_MODELS
+
+from .pagestate import Reader, PageState, SimFitPageState, CUSTOM_MODEL
+from .BumpsFitting import SasFitness, ParameterExpressions
+from .AbstractFitEngine import FitData1D, FitData2D, Model
+from .models import PLUGIN_NAME_BASE, find_plugins_dir
+from .qsmearing import smear_selection
+
+# Monkey patch SasFitness class with plotter
+def sasfitness_plot(self, view='log'):
+    data, theory, resid = self.data.sas_data, self.theory(), self.residuals()
+    plot_theory(data, theory, resid, view)
+SasFitness.plot = sasfitness_plot
+
+# Use a named tuple for the sasview parameters
+PARAMETER_FIELDS = [
+    "fitted", "name", "value", "plusminus", "uncertainty",
+    "lower", "upper", "units",
+    ]
+SasviewParameter = namedtuple("Parameter", PARAMETER_FIELDS)
+
+class FitState(object):
+    def __init__(self, fitfile):
+        self.fitfile = fitfile
+        self.simfit = None
+        self.fits = []
+
+        reader = Reader(self._add_entry)
+        datasets = reader.read(fitfile)
+        self._set_constraints()
+        #print("loaded", datasets)
+
+    def _add_entry(self, state=None, datainfo=None, format=None):
+        """
+        Handed to the reader to receive and accumulate the loaded state objects.
+        """
+        # Note: datainfo is in state.data; format=.svs means reset fit panels
+        if isinstance(state, PageState):
+            # TODO: shouldn't the update be part of the load?
+            state._convert_to_sasmodels()
+            self.fits.append(state)
+        elif isinstance(state, SimFitPageState):
+            self.simfit = state
+        else:
+            # ignore empty fit info
+            pass
+
+    def __str__(self):
+        return '<SasFit %s>'%self.fitfile
+
+    def show(self):
+        """
+        Summarize the fit pages in the state object.
+        """
+        for k, fit in enumerate(self.fits):
```

Comment:
The following loops contain some repetition that can be removed. It is not critical though

Reply:
Stuff 81-87/91-97 into a helper function?